### DISCUSSION TITLE
Are stop visiting goals for multiple stop types possible?

### DISCUSSION THREAD

**Discussion by hklarner** [google/or-tools#3545] [Routing (and legacy CP) questions]
### Problem statement
I have a regular vrp without additional dimensions. It's just a bunch of vehicles the need to visit stops. Vehicles can start and end anywhere. All stops are optional and have a default drop penalty. There are two types of stops, let's call them **red and blue** and in my problems there are different amounts of each available, for example 100 red and 20 blue. Here is my problem: I have visiting goals like "visit at least 10 red and 10 blue stops". The goals don't have to be met precisely.

For the sake of this discussion you can assume that all stop are randomly and evenly distributed on some area. So it looks like white noise on TV, just with red and blue dots (and the ratio dots is not 1:1 but skewed to 1:5, say).

### Solution attempt
My attempt to solve this was to **only work with drop penalties**. The strategy is to define one drop penalty for all red stops and another drop penalty for all blue stops. The idea was to choose the penalties in such a way that I can control the likelihood with which a red stop is visited and the likelihood with which a blue stop is visited.

### Challenges
I can't seem to do the math for computing these penalties correctly. The penalties I compute always result in "bang-bang" solutions, so either all vehicles do all red stops or all vehicles do all blue stops, depending on which penalty is higher.

So I am wondering if there is even a different approach to this problem. I think it should still be a soft constraint in the sense that if the goals are 100/20 then even 112/17 is good.

**Comment by Mizux** [comment]
Did you try to use:
https://github.com/google/or-tools/blob/39a1d96a59dd8fe3d5f6bacfbe14821d2b3d71b7/ortools/constraint_solver/routing.h#L2888-L2898

**Reply by hklarner** [reply]
I haven't. But oof, I wouldn't know how to use this for my original task.
- which nodes / indices should I set this constraint to? to all of them? and what dimension? And what should the lower bound be?
- i don't know how much of one stop type should be serviced already at a given stop, right?

**Comment by jmarca** [comment]
I feel like I already offered my thoughts on this one, but maybe not.

Anyway, I would do two new dimensions, one for red, one for blue.
Each red offers 1 in red dim, 0 in blue dimension.  Vice versa for
blue stops.

Each vehicle is unlimited in both dimensions.

Then at the last stop, use the soft lower bound constraint on each
dimension.

Lower bounds are a problem if they are hard.  If they are not hard,
then the soft lower bound is excellent.


On Sat, Nov 12, 2022 at 09:13:30AM -0800, hklarner wrote:
> I haven't. But oof, I wouldn't know how to use this for my original task.
> - which nodes / indices should I set this constraint to? to all of them? and what dimension? And what should the lower bound be?
> - i don't know how much of one stop type should be serviced already at a given stop, right?
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3545#discussioncomment-4124395
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by hklarner** [reply]
Yes James, I had posted this problem before.

Oh, wow. That is some news! I have tried the hard constraint **on the sum of all last stops,** but got only ROUTING_FAIL_TIMEOUTs.

The things is that this is a constraint that is supposed to act on the vehicle fleet, not on individual vehicles. It is totally fine and desired that some vehicles only do blue stops (if they are in an area where there are mostly blue stops, for example). This could be compensated for by other vehicles doing more red stops. The task is to control to total number of blue and red stops visited by the whole fleet.

So I think right there the `SetCumulVarSoftLowerBound` is limited, since I can't define a soft lower bound over the sum of all last stops. :thinking:

**Comment by jmarca** [comment]
Rereading your original question, you can also use soft upper bound at
the same time, to try to bracket the result at exactly the desired
number of each dimension.

The soft lower bound penalizes the objective if you go under the
target, but then applies zero if you go over.  The soft upper bound
does the opposite, penalizing going over.

So if you want to try to get all vehicles as close to your target
values in all dimensions as possible, you can use both at the same
time.

Hope that helps,

james

On Sat, Nov 12, 2022 at 09:13:30AM -0800, hklarner wrote:
> I haven't. But oof, I wouldn't know how to use this for my original task.
> - which nodes / indices should I set this constraint to? to all of them? and what dimension? And what should the lower bound be?
> - i don't know how much of one stop type should be serviced already at a given stop, right?
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3545#discussioncomment-4124395
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by hklarner** [reply]
The soft upper bound together with the soft lower bound sounds like a great idea :+1: but I still don't see how I can use these methods to control the visited stops for the whole fleet and not for individual vehicles.

**Comment by jmarca** [comment]
Ah I missed that feature. 

I’d start by making new variables that just track the total of each color served. That’s the first step. Then try to figure out a way to constrain it as needed. 

The trouble with the routing solver is that the objective is vehicle/route centric. It is hard to directly modify it.  Well there may be a way but I do not know of it.  So I think you’ll have to figure out some trick to make it work. 

James
